File: fetcher-go/internal/gitlab/gitlab.go
// Package gitlab: GitLab REST API クライアント (必要な分だけ)。
// Rust 版 (fetcher-rs/src/gitlab.rs) と同じエンドポイント・同じレート制御 (1 req/sec)。
//
// プロキシは http.Transport がデフォルトで HTTP_PROXY / HTTPS_PROXY を見るので追加対応不要。
// TLS は crypto/tls (純 Go) で、Windows の schannel は経由しない。
package gitlab

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sync"
	"time"

	"repo-tracker/internal/model"
)

// Client はレート制御と http.Client を持つ。
type Client struct {
	http  *http.Client
	token string

	// 1 fetch / 1 sec を守るためのロック。
	mu    sync.Mutex
	last  time.Time
	delay time.Duration
}

func New(token string) *Client {
	return &Client{
		http:  &http.Client{Timeout: 60 * time.Second},
		token: token,
		delay: time.Second,
	}
}

// sendJSON: GET → JSON デコード。1 req/sec のレート制限を強制する。
func (c *Client) sendJSON(rawURL string, out interface{}) error {
	c.wait()
	req, err := http.NewRequest(http.MethodGet, rawURL, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "repo-tracker/0.1")
	if c.token != "" {
		req.Header.Set("PRIVATE-TOKEN", c.token)
	}
	resp, err := c.http.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}
	return json.Unmarshal(body, out)
}

// 直前のリクエストから delay 経過するまで sleep。
func (c *Client) wait() {
	c.mu.Lock()
	defer c.mu.Unlock()
	if !c.last.IsZero() {
		elapsed := time.Since(c.last)
		if elapsed < c.delay {
			time.Sleep(c.delay - elapsed)
		}
	}
	c.last = time.Now()
}

// ---------- /projects/:id ----------

type ProjectMeta struct {
	ID                uint64 `json:"id"`
	Name              string `json:"name"`
	PathWithNamespace string `json:"path_with_namespace"`
	WebURL            string `json:"web_url"`
	DefaultBranch     string `json:"default_branch"`
}

func (c *Client) FetchProjectMeta(baseURL, project string) (*ProjectMeta, error) {
	u := fmt.Sprintf("%s/api/v4/projects/%s",
		trimTrailingSlash(baseURL),
		url.PathEscape(project),
	)
	var m ProjectMeta
	if err := c.sendJSON(u, &m); err != nil {
		return nil, err
	}
	return &m, nil
}

// ---------- /projects/:id/repository/tree (walked) ----------

type treeEntry struct {
	Name string `json:"name"`
	Type string `json:"type"` // "blob" or "tree"
	Path string `json:"path"`
}

// ListTrackedFiles: tree API を nest 回まで再帰して、targets に含まれる blob 名のみ返す。
func (c *Client) ListTrackedFiles(
	baseURL, project string,
	nest uint32,
	targets []string,
) ([]model.RepoFileRef, error) {
	var out []model.RepoFileRef
	if err := c.visitDir(baseURL, project, "", nest, targets, &out); err != nil {
		return nil, err
	}
	return out, nil
}

func (c *Client) visitDir(
	baseURL, project, path string,
	remaining uint32,
	targets []string,
	out *[]model.RepoFileRef,
) error {
	entries, err := c.fetchTree(baseURL, project, path)
	if err != nil {
		return err
	}
	for _, e := range entries {
		switch {
		case e.Type == "blob" && contains(targets, e.Name):
			if kind, ok := model.ClassifyFileName(e.Name); ok {
				*out = append(*out, model.RepoFileRef{Type: kind, Path: e.Path})
			}
		case e.Type == "tree" && remaining > 0:
			if err := c.visitDir(baseURL, project, e.Path, remaining-1, targets, out); err != nil {
				return err
			}
		}
	}
	return nil
}

func (c *Client) fetchTree(baseURL, project, path string) ([]treeEntry, error) {
	u := fmt.Sprintf("%s/api/v4/projects/%s/repository/tree?per_page=100",
		trimTrailingSlash(baseURL), url.PathEscape(project))
	if path != "" {
		u += "&path=" + url.QueryEscape(path)
	}
	var entries []treeEntry
	if err := c.sendJSON(u, &entries); err != nil {
		return nil, err
	}
	return entries, nil
}

// ---------- /projects/:id/repository/files/:path ----------

type RawFile struct {
	BlobSHA string
	Size    uint64
	Raw     string
}

type fileResp struct {
	Size     uint64 `json:"size"`
	Encoding string `json:"encoding"`
	BlobID   string `json:"blob_id"`
	Content  string `json:"content"`
}

func (c *Client) FetchFile(baseURL, project, ref, path string) (*RawFile, error) {
	u := fmt.Sprintf("%s/api/v4/projects/%s/repository/files/%s?ref=%s",
		trimTrailingSlash(baseURL),
		url.PathEscape(project),
		url.PathEscape(path),
		url.QueryEscape(ref),
	)
	var r fileResp
	if err := c.sendJSON(u, &r); err != nil {
		return nil, err
	}
	if r.Encoding != "base64" {
		return nil, fmt.Errorf("unexpected encoding: %s", r.Encoding)
	}
	decoded, err := base64.StdEncoding.DecodeString(r.Content)
	if err != nil {
		return nil, fmt.Errorf("base64 decode: %w", err)
	}
	return &RawFile{
		BlobSHA: r.BlobID,
		Size:    r.Size,
		Raw:     string(decoded),
	}, nil
}

// ---------- helpers ----------

func trimTrailingSlash(s string) string {
	for len(s) > 0 && s[len(s)-1] == '/' {
		s = s[:len(s)-1]
	}
	return s
}

func contains(xs []string, v string) bool {
	for _, x := range xs {
		if x == v {
			return true
		}
	}
	return false
}

File: fetcher-rs/src/gitlab.rs
// GitLab API client (just the bits we need).
//
// All requests throttle to 1 fetch / sec inside send_json, per DESIGN.md.

use crate::model::{FileKind, RepoFileRef};
use reqwest::blocking::Client;
use reqwest::header;
use serde::Deserialize;
use std::error::Error;
use std::thread;
use std::time::Duration;

pub fn build_client() -> Result<Client, Box<dyn Error>> {
    // 明示的に rustls を選ぶ。reqwest がデフォルトで OS の TLS スタックを掴むのを
    // 防ぎたい (Windows の schannel + 社内プロキシで CRL/OCSP 取得が失敗するため)。
    Ok(Client::builder()
        .use_rustls_tls()
        .user_agent("repo-tracker/0.1")
        .build()?)
}

// ---------- /projects/:id ----------

#[derive(Debug, Deserialize)]
pub struct ProjectMeta {
    pub id: u64,
    pub name: String,
    pub path_with_namespace: String,
    pub web_url: String,
    pub default_branch: String,
}

pub fn fetch_project_meta(
    client: &Client,
    base_url: &str,
    project: &str,
    token: Option<&str>,
) -> Result<ProjectMeta, Box<dyn Error>> {
    let url = format!(
        "{}/api/v4/projects/{}",
        base_url.trim_end_matches('/'),
        urlencoding::encode(project),
    );
    send_json(client, &url, token)
}

// ---------- /projects/:id/repository/files/:path ----------

pub struct RawFile {
    pub blob_sha: String,
    pub size: u64,
    pub raw: String,
}

pub fn fetch_file(
    client: &Client,
    base_url: &str,
    project: &str,
    token: Option<&str>,
    ref_: &str,
    path: &str,
) -> Result<RawFile, Box<dyn Error>> {
    #[derive(Deserialize)]
    struct FileResp {
        size: u64,
        encoding: String,
        blob_id: String,
        content: String,
    }

    let url = format!(
        "{}/api/v4/projects/{}/repository/files/{}?ref={}",
        base_url.trim_end_matches('/'),
        urlencoding::encode(project),
        urlencoding::encode(path),
        urlencoding::encode(ref_),
    );
    let resp: FileResp = send_json(client, &url, token)?;

    if resp.encoding != "base64" {
        return Err(format!("unexpected encoding: {}", resp.encoding).into());
    }

    use base64::Engine;
    let bytes = base64::engine::general_purpose::STANDARD.decode(&resp.content)?;
    let raw = String::from_utf8(bytes)?;

    Ok(RawFile {
        blob_sha: resp.blob_id,
        size: resp.size,
        raw,
    })
}

// ---------- /projects/:id/repository/tree (walked) ----------

#[derive(Debug, Deserialize)]
struct TreeEntry {
    name: String,
    #[serde(rename = "type")]
    kind: String, // "blob" or "tree"
    path: String,
}

pub fn list_tracked_files(
    client: &Client,
    base_url: &str,
    project: &str,
    token: Option<&str>,
    nest: u32,
    targets: &[&str],
) -> Result<Vec<RepoFileRef>, Box<dyn Error>> {
    let mut results = Vec::new();
    visit_dir(
        client, base_url, project, token, "", nest, targets, &mut results,
    )?;
    Ok(results)
}

fn visit_dir(
    client: &Client,
    base_url: &str,
    project: &str,
    token: Option<&str>,
    path: &str,
    remaining: u32,
    targets: &[&str],
    out: &mut Vec<RepoFileRef>,
) -> Result<(), Box<dyn Error>> {
    let entries = fetch_tree(client, base_url, project, token, path)?;

    for e in entries {
        // println!("  {:?}", e);
        match e.kind.as_str() {
            "blob" if targets.iter().any(|t| *t == e.name) => {
                if let Some(kind) = classify(&e.name) {
                    out.push(RepoFileRef { kind, path: e.path });
                }
            }
            "tree" if remaining > 0 => {
                visit_dir(
                    client,
                    base_url,
                    project,
                    token,
                    &e.path,
                    remaining - 1,
                    targets,
                    out,
                )?;
            }
            _ => {}
        }
    }
    Ok(())
}

fn fetch_tree(
    client: &Client,
    base_url: &str,
    project: &str,
    token: Option<&str>,
    path: &str,
) -> Result<Vec<TreeEntry>, Box<dyn Error>> {
    let path_query = if path.is_empty() {
        String::new()
    } else {
        format!("&path={}", urlencoding::encode(path))
    };
    let url = format!(
        "{}/api/v4/projects/{}/repository/tree?per_page=100{}",
        base_url.trim_end_matches('/'),
        urlencoding::encode(project),
        path_query,
    );
    send_json(client, &url, token)
}

fn classify(name: &str) -> Option<FileKind> {
    match name {
        "package.json" => Some(FileKind::PackageJson),
        "pom.xml" => Some(FileKind::PomXml),
        ".gitlab-ci.yml" => Some(FileKind::GitlabCi),
        _ => None,
    }
}

// ---------- shared GET → JSON ----------

fn send_json<T: serde::de::DeserializeOwned>(
    client: &Client,
    url: &str,
    token: Option<&str>,
) -> Result<T, Box<dyn Error>> {
    let mut req = client.get(url).header(header::ACCEPT, "application/json");
    if let Some(t) = token {
        req = req.header("PRIVATE-TOKEN", t);
    }
    let resp = req.send()?;
    let status = resp.status();
    if !status.is_success() {
        let body = resp.text().unwrap_or_default();
        return Err(format!("HTTP {status}: {body}").into());
    }
    let data: T = resp.json()?;
    // Per design: throttle to 1 fetch / sec.
    thread::sleep(Duration::from_secs(1));
    Ok(data)
}
